### ISSUE TITLE
How to trigger a tool recalculation after 'addToolState' for RectangleRoiTool?

### ISSUE THREAD

**Issue by pzehle** [cornerstonejs/cornerstoneTools#877]
### Prerequisites

* [x] Are you running the latest version?
* [x] Are you reporting to the correct repository?
* [x] Did you perform a cursory search?

For more information, see the `CONTRIBUTING` guide.

### Description

I am working currently on a tool that has to calculate a volume, this means, I draw an area using the `RectangleRoiTool`, and then I have to replicate that area on X number of images. I am doing this by using `addToolState` on each image, using the `toolState` from the area I created. The problem is, when the script scrolls to the next image and adds the tool state, it doesn't recalculate the Mean and StdDev of the current image, but it is saving the data from the created area. Is there an option to recalculate the state once it is added? Specifically for the `RectangleRoiTool`.

Thanks!

### Steps to Reproduce

1. Draw a rectangle using `RectangleRoiTool`.
2. Save the tool state on a variable.
3. Add the saved tool state on any other image. The calculations are still the ones from the variable, not the ones that should be from this new image.

**Expected behavior:** Once the tool state is added, it should recalculate the information from this new image.

**Actual behavior:** After setting the tool state, the information from the saved state is static, no recalculation is being triggered.

**Comment by dannyrb** [comment]
@pzehle, I don't believe this feature exists. I would be open to a PR that updates the `BaseAnnotationTool` to accomplish this. Something like...

`forceRecalc` that has a default implementation of setting the tool's measurement as "dirty" and triggering a re-render. This is definitely up for discussion. @JamesAPetts may have other thoughts.

Small note. If you're calculating volume, you may not want to sum area on X slides. You might be better off calculating the volume of your cube, then using that cube's dimensions to determine how to annotate each slice.

**Comment by JamesAPetts** [comment]
This is an interesting problem and one I've thought of for a while. Shared data between annotations are what modules are built for, and this is a perfect use case for that.

I would just give the rects a UID, and register them to a ledger in a module. In the rects `addToolState` function I'd add a call to the module to invalidate all annotations registered in the ledger.

**Comment by JamesAPetts** [comment]
You could also save the volume as a module variable, and have a getter in your render function for the toolState that fetches this, so you only have to keep track of one copy of it.

**Comment by pzehle** [comment]
@dannyrb Thanks for the response. I will think of an implementation for this and let you know of my thoughts. Maybe we can include this as a PR. I will let you know. As of the volume note, thanks, I will keep in mind that, but I actually need the density of the HU of the cube, which I think is a lot more important for my development, as we are trying to implement some ML and tumour detection.

@JamesAPetts Thanks for the response as well. I will think of a solution for this. My problem is that I have to deliver a quick example this Friday, no time to develop complex stuff. You think, if after adding the tool state, I trigger a click on one of the handles to update the calculations, which manually works, is a good idea?

I am up to quick solutions if you have them, if not, I will think of something for the library and instead I am going for something temporary. Thanks both!

**Comment by pzehle** [comment]
@dannyrb, @JamesAPetts hey, no luck with the triggered click, don't know why but it doesn't work. Digging further in the code, I found that the `RectangleRoiTool.js` source file has a function called `_updateCachedStats`. Isn't it possible to call that function from within my code?

Thanks!

**Comment by dannyrb** [comment]
@pzehle, probably, but it's not "public" and could change.

If you need something quick, we could:

- Restore state with no calculated value
- Put in a PR where, if the value is undefined, it should be recalculated in the render loop

If you need to force a re-render for all elements, you can do something like:

```js
import { updateImage } from 'cornerstone-core';
import { store } from 'cornerstone-tools';

store.state.enabledElements.forEach(enabledElement =>
    updateImage(enabledElement );
  );
```

**Comment by pzehle** [comment]
@dannyrb thanks for the reply. I've tried this, and sadly it updates the image but doesn't recalculate the information within the area, it just stays the same. I am bit lost here.

**Comment by dannyrb** [comment]
@pzehle, right. You would also need a PR that changes this line:

https://github.com/cornerstonejs/cornerstoneTools/blob/master/src/tools/annotation/RectangleRoiTool.js#L191

```
if (data.invalidated === true || !data.cachedStats) {
  ...
}
```

**Comment by pzehle** [comment]
@dannyrb thanks, you are always helpful. I am still lost, but I think you gave me a place to start. Thanks!

**Comment by dannyrb** [comment]
How are you `console.log()`-ing? I could be wrong, but for objects I think the value is evaluated when you expand the object in the console, not at the time of logging.

When you restore state, are you making sure `invalidated` is `true` and `cachedStats` is `undefined`?

**Comment by pzehle** [comment]
I am logging directly on my code, and you are right, but that happens **only** some times. This has been a difficult development as my code has so many promises and async stuff, that I think most of the time is running so fast that values are not being properly updated. Also, I forgot to set `cachedStats` as `undefined`, and I just did that and it crashed, because of the async stuff. I will look into it and let you know. Thanks again.

**Comment by pzehle** [comment]
Hello @dannyrb, as always, thank you very much.

I managed to accomplish this setting  `invalidated` as `true` and `cachedStats` as `undefined` as you suggested. I was having the problem of the tool still not updating, but I've restructured my process with a lot of promises, and have a `timeout` to wait for the tool to recalculate, this was the trick.

I will work on the PR you suggested to add this functionality. Thanks!

**Comment by dannyrb** [comment]
👋 Hi @pzehle, I'm going to close this issue for now. It sounds like there's an "okay" workaround that helped you accomplish your goal.

If you think there's an improvement we can make to simplify this, please feel free to create a separate discussion issue for the change.

**Comment by pzehle** [comment]
Hi @dannyrb. For now is OK. I will work in the future on a PR to have an automated method to accomplish this without any of this processes. I will let you know. Thanks!

**Comment by JamesAPetts** [comment]
@pzehle @dannyrb 

Hey y'all, just a heads up, I'm going to build an API to help with stuff like this, for plugin development for larger/3D tools.

On the drawing board:
```js
// Returns toolState for each image in the stack with toolState.
// If tool is specified, only that tool's state.
getToolStateForStack(
  element,
  [tool]
)

// Invalidates any toolState in the stack that has an `invalidated` method.
invalidateToolStateForStack(
  element,
  [tool]
)

// Cleanly merges toolState to globalToolState.
mergeToolStateToGlobalToolState(
  toolState
)

// clears toolstate for all imageIds in the stack.
// If a tool is specified, then only clear that tool.
clearToolStateForStack(
  element,
  [tool]
)
```

Probably expose this stuff under something like:

```js
const {
  getToolStateForStack,
  invalidateToolStateForStack,
  mergeToolStateToGlobalToolState,
  clearToolStateForStack
} = cornerstoneTools.import('utils/stackToolStateUtils');
```

Let me know if you think of any additional functionality that'd be useful for your usecase.

**Comment by pzehle** [comment]
Hi @JamesAPetts, I think it is an excellent idea, mainly because right now it is complicated to achieve what I wanted, though it was relatively easy to accomplish, I think this would resolve a lot of issues.

**Comment by JamesAPetts** [comment]
Addressed in #889 .

**Comment by supasus** [comment]
I have a tool that changes colPixelSpacing and rowPixelSpacing it works fine on tools that are used after the calibration, however any measurements that were there before do not get updated unless I touch them. How can I force an update?
